doxygen/contributor_help_pages/common_pitfalls.md

```diff
@@ -0,0 +1,239 @@
+else. This is convenient because it works on non-Eigen types as well
+(as compared to the built in `.eval()` member function on Eigen
+types).
+
+The implementation of `make_holder` is
+[here](https://github.com/stan-dev/math/blob/develop/stan/math/prim/meta/holder.hpp).
+
+### Move Semantics
+
+In general, Stan math does not use move semantics very often. This is because of our arena allocator. Move semantics generally work as
+
+```cpp
+
+class my_big_type {
+  double* m_data;
+
+  explicit my_big_type(size_t N) m_data(static_cast<double*>(malloc(N))) {}
+  // Standard copy constructor
+  my_big_type(const my_big_type& other) m_data(static_cast<double*>(malloc(N))) {
+    std::copy(other.m_data, this->m_data);
+  }
+  // Standard move constructor
+  my_big_type(my_big_type&& other) m_data(other.m_data) {
+    other.m_data = nullptr;
+  }
+};
+```
+
+We can see in the above that the standard style of a move (the constructor taking an rvalue reference) is to copy the pointer and then null out the original pointer. But in Stan, particularly for reverse mode, we need to keep memory around even if  it's only a temporary for when we call the gradient calculations in the reverse pass. And since memory for reverse mode is stored in our arena allocator no copying happens in the first place.
+
+When working with arithmetic types, keep in mind that moving Scalars is often less optimal than simply taking their copy. For instance, Stan's `var` type is a PIMPL implimentation, so it simply holds a pointer of size 8 bytes. A `double` is also 8 bytes which just so happens to fit exactly in a [word](https://en.wikipedia.org/wiki/Word_(computer_architecture)) of most modern CPUs. While a reference to a double is also 8 bytes, unless the function is inlined by the compiler, the computer will have to place the reference into cache, then go fetch the value that is being referenced which now takes up two words instead of one!
+
+The general rules to follow for passing values to a function are:
+
+1. If a type is smaller than two words (16 bytes), pass it by value
+2. If you are writing a function for reverse mode, pass values by `const&`
+3. In prim, if you are confident and working with larger types, use perfect forwarding to pass values that can be moved from. Otherwise simply pass values by `const&`.
+
+### Copying non-arena variables to lambdas used in the reverse pass
+
+### Returning arena types
+
+### `.val()` vs. `.val_op()`
+
+### Returning expressions
+
+### Const correctness, reverse mode autodiff, and arena types
+
+## Handy tricks
+
+### `make_callback_ptr`
+
+### `forward_as`
+
+### Nested Stacks
+
+### Scoped Stacks
+
+As an aside, in order to get the actual vector result, Eigen defines the
+`.eval()` operator, so `c.eval()` would return an actual `Eigen::VectorXd`.
+Similarly, Math defines the `eval` function which will evaluate an Eigen
+expression and forward anything that is not an Eigen expression.
```

Comment:
@wds15 would you have time to write about Stan's threading scheme for the allocators? I think that would be best to put here